src/support/logger/logger.c

```diff
@@ -11,27 +16,69 @@ void logger_setDefault(Logger* logger) {
 
 Logger* logger_getDefault() { return defaultLogger; }
 
-static GLogLevelFlags _gloglevelflags(LogLevel level) {
-    switch (level) {
-        case LOGLEVEL_CRITICAL: return G_LOG_LEVEL_CRITICAL;
-        case LOGLEVEL_WARNING: return G_LOG_LEVEL_WARNING;
-        case LOGLEVEL_MESSAGE: return G_LOG_LEVEL_MESSAGE;
-        case LOGLEVEL_INFO: return G_LOG_LEVEL_INFO;
-        case LOGLEVEL_DEBUG: return G_LOG_LEVEL_DEBUG;
-        case LOGLEVEL_UNSET:
-        case LOGLEVEL_ERROR:
-        default: return G_LOG_LEVEL_ERROR;
+// Process start time, initialized explicitly or on first use.
+static pthread_mutex_t _start_time_mutex = PTHREAD_MUTEX_INITIALIZER;
+static bool _start_time_initd = false;
+static int64_t _monotonic_start_time_micros;
+
+int64_t logger_now_micros() {
+    return g_get_monotonic_time();
+}
+
+int64_t logger_get_global_start_time_micros() {
+    pthread_mutex_lock(&_start_time_mutex);
+    if (!_start_time_initd) {
+        // TODO: Parse out of /proc/$$/stat instead to get the true process
+        // start time?
+        _monotonic_start_time_micros = logger_now_micros();
+        _start_time_initd = true;
     }
+    int64_t t = _monotonic_start_time_micros;
+    pthread_mutex_unlock(&_start_time_mutex);
+    return t;
+}
```

Comment:
Should we really lock a mutex on every single log message, just to be able to set the start time? I think we can avoid it if we use atomic operations (`#include <stdatomic.h>` in the C11 standard).

Assuming you have defined `volatile atomic_flag _start_time_initd = ATOMIC_FLAG_INIT;`, then something like:

```
int64_t logger_get_global_start_time_micros() {
    if(!atomic_flag_test_and_set(&_start_time_initd)) {
        logger_set_global_start_time_micros(logger_now_micros());
    }
    return _monotonic_start_time_micros;
}
```

See also: https://en.cppreference.com/w/c/atomic

Reply:
Thinking about this some more, we probably don't need to call the setter after any additional threads are started. I'll change it and note that restriction.

Reply:
> More generally, grabbing a mutex is cheap anyway as long there's not actually any contention; in the common case we're only executing a couple instructions with the mutex held so there's unlikely to be any contention.

It becomes more likely as the number of Shadow worker threads increases, and especially as we increase the logging verbosity. I currently run some experiments on an 80 hyperthread machine, and all of them could be trying to log at the same time.

> I'd originally used pthread_once, which probably does something like this under the hood, but changed it when I realized I needed a setter for use inside the shim. Using atomic/pthread once primitives won't address parallel access via the setter.

Is your argument that using `pthread_once` in the getter and keeping the mutex for the setter is too much added complexity? It does seem worth it to me in order to **guarantee** that contention doesn't reduce performance.

Reply:
(Whoops, it looks like your comment came in as I was typing mine and I either missed it or Github didn't load it until I just refreshed the page.)

Reply:
Actually just using an atomic wouldn't give us safety since if the timer isn't initialized explicitly, and two threads enter this code simultaneously, the second thread needs to wait for the first thread to finish initializing.

We can handle that case with pthread_once instead (with the caveat that the setter is no longer thread safe).

> pthread_once in the getter and keeping the mutex for the setter is too much added complexity?

I don't think that would quite do what we want. If we want the setter to be thread-safe, then the getter would also need to lock the mutex.

I think the possibilities here are:

1. Just use a mutex as the code is now. This is the simplest, and IMO unlikely to be a performance issue even in the 80 hyperthread case due to only executing a few instructions while holding the mutex. I don't think it's worth spending the time trying to validate that now though, so alternatively:
2. Use [pthread read/write locks](https://linux.die.net/man/3/pthread_rwlock_init), which allow parallel "readers" under the lock. I think this would address your concern *and* let us have a thread-safe setter. What I've "heard" though is that the extra complexity in read/write locks actually make them less efficient in cases like this one where the lock isn't held anyway, though.
3. Use pthread_once and declare that the setter is not thread safe. I think we can agree that this is the most performant option, and that we don't actually need the setter to be thread safe?

I'm leaning towards 3 and will implement it shortly; LMK if you disagree :)

Reply:
I don't know much about mutex performance, so I'm willing to defer to your judgement and go with option 1.

But if you are now convinced about option 3, that's fine too :)

Reply:
Unless we have a reason to make the setter thread safe, I think no reason not to go with #3 :)

Reply:
For future reference: Even without contention, lots of atomic operations (including locking) can slow compute throughput down, especially on large multi-CPU machines, because every call to an atomic operation requires cache coherence across all cores, even when there's no data dependence.

Reply:
I wonder if it's possible to use a constructor to set the init time and then avoid the `pthread_once` call? Something like `__attribute__((constructor(PRIORITY))) static void _init_func() {}`.

Not sure if it's worth it at this point, but wanted to mention it as another possible option here.

Reply:
> I wonder if it's possible to use a constructor to set the init time and then avoid the pthread_once call? Something like __attribute__((constructor(PRIORITY))) static void _init_func() {}.

It's possible, but using constructors can be a bit frought to ensure that they get called in the correct order.

Btw re @jtracey's comment I looked a bit closer at the disassembly for pthread_once and  that once initialization has happened there's no atomic operation. It just compares against a plain bool first and only if that hasn't been set yet gets into the synchronization bits. i.e. after initialization the pthread_once overhead should be negligible - just a function call and a bool comparison.

Reply:
I was actually thinking about going the other direction and switching over some of the other constructor usage in the shim to use pthread_once as well. I think the only constructor we need is the main shim startup "entry point"; it's probably safer and clearer to use lazy initialization anywhere else we need something init'd.

Reply:
Thanks for looking into `pthread_once` - it seems like the best way to go!